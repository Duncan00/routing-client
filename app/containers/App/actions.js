import {
  LOAD_ROUTE,
  LOAD_ROUTE_SUCCESS,
  LOAD_ROUTE_ERROR,
  CREATE_ROUTE,
  CREATE_ROUTE_SUCCESS,
  CREATE_ROUTE_ERROR,
  RESET,
} from './constants';

export function loadRoute() {
  return {
    type: LOAD_ROUTE,
  };
}

export function routeLoaded(data) {
  return {
    type: LOAD_ROUTE_SUCCESS,
    data,
  };
}

export function routeLoadingError(error) {
  return {
    type: LOAD_ROUTE_ERROR,
    error,
  };
}

export function createRoute(startingLocation, dropOffPoint) {
  return {
    type: CREATE_ROUTE,
    startingLocation,
    dropOffPoint,
  };
}

export function routeCreated(data) {
  return {
    type: CREATE_ROUTE_SUCCESS,
    data,
  };
}

export function routeCreatingError(error) {
  return {
    type: CREATE_ROUTE_ERROR,
    error,
  };
}

export function reset() {
  return {
    type: RESET,
  };
}
