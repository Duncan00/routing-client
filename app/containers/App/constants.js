/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_ROUTE = 'routing-client/App/LOAD_ROUTE';
export const LOAD_ROUTE_SUCCESS = 'routing-client/App/LOAD_ROUTE_SUCCESS';
export const LOAD_ROUTE_ERROR = 'routing-client/App/LOAD_ROUTE_ERROR';

export const CREATE_ROUTE = 'routing-client/App/CREATE_ROUTE';
export const CREATE_ROUTE_SUCCESS = 'routing-client/App/CREATE_ROUTE_SUCCESS';
export const CREATE_ROUTE_ERROR = 'routing-client/App/CREATE_ROUTE_ERROR';

export const RESET = 'routing-client/App/RESET';
