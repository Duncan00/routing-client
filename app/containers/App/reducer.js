import produce from 'immer';
import { pick } from 'lodash';
import {
  LOAD_ROUTE,
  LOAD_ROUTE_SUCCESS,
  LOAD_ROUTE_ERROR,
  CREATE_ROUTE,
  CREATE_ROUTE_SUCCESS,
  CREATE_ROUTE_ERROR,
  RESET,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  api: process.env.API,
  user_data: {
    route_token: '',
    route: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  // eslint-disable-next-line consistent-return
  produce(state, draft => {
    switch (action.type) {
      case LOAD_ROUTE:
        draft.loading = true;
        draft.error = false;
        draft.user_data.route = false;
        break;

      case LOAD_ROUTE_SUCCESS:
        draft.user_data.route = pick(action.data, [
          'path',
          'total_distance',
          'total_time',
        ]);
        draft.loading = false;
        break;

      case LOAD_ROUTE_ERROR:
        draft.error = action.error.message;
        draft.loading = false;
        break;

      case CREATE_ROUTE:
        draft.loading = true;
        draft.error = false;
        draft.user_data.route_token = '';
        break;

      case CREATE_ROUTE_SUCCESS: {
        draft.loading = false;
        draft.user_data.route_token = action.data.token;
        break;
      }

      case CREATE_ROUTE_ERROR:
        draft.loading = false;
        draft.error = action.error.message;
        break;

      case RESET:
        return initialState;
    }
  });

export default appReducer;
