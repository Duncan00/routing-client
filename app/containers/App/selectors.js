import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const makeSelectApi = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.api,
  );

const makeSelectRouteToken = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.user_data.route_token,
  );

const makeSelectRoute = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.user_data.route,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

export {
  makeSelectApi,
  makeSelectRouteToken,
  makeSelectRoute,
  makeSelectError,
  makeSelectLoading,
};
