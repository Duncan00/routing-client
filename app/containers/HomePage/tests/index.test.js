import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { HomePage, mapDispatchToProps } from '../index';
import { createRoute } from '../../App/actions';
import configureStore from '../../../configureStore';
import setGlobalGoogleMock from '../../../../internals/testing/setGlobalGoogleMock';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    setGlobalGoogleMock();

    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <HomePage
            error={false}
            loading={false}
            onSubmitForm={() => {}}
            onResetForm={() => {}}
            route={{
              total_distance: 1000,
              total_time: 100,
              path: [
                ['22.372081', '114.107877'],
                ['22.326442', '114.167811'],
                ['22.284419', '114.159510'],
              ],
            }}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  describe('mapDispatchToProps', () => {
    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch createRoute when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const values = {
          starting_location: 'starting_location',
          drop_off_point: 'drop_off_point',
        };
        const setSubmitting = () => {};
        result.onSubmitForm(values, { setSubmitting });
        expect(dispatch).toHaveBeenCalledWith(
          createRoute(values.starting_location, values.drop_off_point),
        );
      });
    });
  });
});
