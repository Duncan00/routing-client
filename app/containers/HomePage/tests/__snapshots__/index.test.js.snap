// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
.c6 {
  color: red;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c1 {
  padding: 30px 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  max-height: 360px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c8 {
  width: 100px;
}

.c5 {
  padding: 0;
  border: none;
  background: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c4 {
  width: 90%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100vh;
}

@media only screen and (max-width:600px) {
  .c1 {
    width: 50%;
  }
}

@media only screen and (min-width:600px) {
  .c1 {
    width: 360px;
  }
}

<div
  class="c0"
>
  <form
    action="#"
    class="c1"
  >
    <div
      class="c2"
    >
      <label
        for="starting_location"
      >
        Starting Location
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="starting_location"
          name="starting_location"
          value="Innocentre, Hong Kong"
        />
        <button
          class="c5"
          type="button"
        >
          ✖
        </button>
      </div>
      <div
        class="c6"
      />
    </div>
    <div
      class="c2"
    >
      <label
        for="drop_off_point"
      >
        Drop Off Point
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="drop_off_point"
          name="drop_off_point"
          value="Hong Kong International Airport Terminal 1"
        />
        <button
          class="c5"
          type="button"
        >
          ✖
        </button>
      </div>
      <div
        class="c6"
      />
    </div>
    <div
      class="c2"
    />
    <div
      class="c2"
    >
      <label>
        total distance: 
        1000
      </label>
      <label>
        total time: 
        100
      </label>
    </div>
    <div
      class="c7"
    >
      <button
        class="c8"
        type="submit"
      >
        Re-Submit
      </button>
      <button
        class="c8"
        type="reset"
      >
        Reset
      </button>
    </div>
  </form>
  <div
    style="width: 100%;"
  >
    <div
      style="height: 100%;"
    />
    <div>
      <div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
